README.md
## Onion Browser

[![Build Status](https://travis-ci.org/mtigas/OnionBrowser.svg?branch=2.X)](https://travis-ci.org/mtigas/OnionBrowser)  
[Official Site][official] | [Support][help] | [Changelog][changelog] | [Donate][donate]  
&copy; 2012-2017 [Mike Tigas][miketigas] ([@mtigas](https://twitter.com/mtigas))

*This is the Onion Browser <strong>2.X branch</strong>, based on [Endless][endless]. This branch is in active development and is probably completely broken from time to time. This version has not yet been released in the App Store.*

**Onion Browser** is a free web browser for iPhone and iPad that encrypts and tunnels web traffic through the [Tor network][tor]. See the [official site][official] for more details and App Store links.

Please see the [LICENSE][license] file for usage and redistribution terms. As of the 2.X (Endless-based) tree, redistribution of this software in binary form, with or without modification, is not permitted. (The previous [1.X tree][1.X] of Onion Browser was available under [a different license](https://github.com/mtigas/OnionBrowser/blob/1.X/LICENSE).)

---

* **OnionBrowser**: 2.0.0-dev (XXXXXXXX.X) - See [official release history][releases] and [changelog][changelog].
* **[Endless][endless]** upstream: 1.6.1
* **[Tor.framework][Tor.framework]**: f4cae85 (Apr 28 2017)
  * **[Tor][tor]**: 0.3.0.6
  * **[libevent][libevent]**: 2.1.8
  * **[OpenSSL][openssl]**: 1.0.2k

[official]: https://mike.tig.as/onionbrowser/
[help]: https://mike.tig.as/onionbrowser/help/
[releases]: https://github.com/mtigas/OnionBrowser/releases
[changelog]: https://raw.github.com/mtigas/OnionBrowser/2.X/CHANGES.txt
[donate]: https://mike.tig.as/onionbrowser/#support-project
[miketigas]: https://mike.tig.as/
[license]: https://github.com/mtigas/OnionBrowser/blob/2.X/LICENSE
[Tor.framework]: https://github.com/iCepa/Tor.framework
[tor]: https://www.torproject.org/
[libevent]: http://libevent.org/
[openssl]: https://www.openssl.org/
[iobfs]: https://github.com/mtigas/iObfs
[endless]: https://github.com/jcs/endless
[1.X]: https://github.com/mtigas/OnionBrowser/tree/1.X

### Notable 2.X Features

The following features are new to Onion Browser, by way of the upstream work on [Endless][endless]:

- Multiple tab support

- Search from URL bar

- Ability to configure security and privacy settings (script blocking, etc) on a per-site basis

- Per-site cookie handling

- [HTTPS Everywhere](https://www.eff.org/HTTPS-EVERYWHERE) support

- HTTP Strict Transport Security (HSTS) support, pre-loaded with the [Chromium ruleset](https://hstspreload.org/)

- Ability to view SSL certificate information, to allow manual verification of SSL certificates

- [1Password extension](https://github.com/AgileBits/onepassword-app-extension)
  support (if 1Password app is installed)


CHANGES.txt
Onion Browser Version History (since 1.2.2)

FUTURE:

===== 2.0 - In Development =====
Features subject to change.

* New design and new codebase, based on the Endless Browser open source
  project by jcs, used with permission; many thanks to jcs for allowing us
  to build off of his work. Major highlights:

  - Tabbed browsing

  - Built-in HTTPS Everywhere ruleset

  - Built-in tracking prevention and ad blocking

  - HTTP Strict Transport Security (HSTS) support

  - SSL certificate viewer

  - Ability to set privacy on a per-site basis

  - 1Password extension support

* New "flat" icon.

? Minimum required iOS version is now iOS 9.0 (was iOS 8.2).

? Data encryption with an app-level password. Supplements current
  encryption done by the iOS operating system, improving protection
  against forensic tools.

===== 1.7.5 - Pending App Store Approval =====
May 21 2017 - Submit Date

* Built-in obfs4 and meek-amazon bridges updated to latest Tor Browser
  Bundle configuration.

* Tor updated to 0.3.0.7, fixing several Tor networking bugs.



CURRENT VERSION:

===== 1.7.4 - In App Store =====
Apr 02 2017 - Submit Date
Apr 04 2017 - App Store Availability

* Built-in bridges updated to latest Tor Browser Bundle configuration.
  (Fixes non-working meek-azure.)

* Tor updated to 0.3.0.4-rc.

* Updated user-agent spoofing strings.


PREVIOUS VERSIONS:

===== 1.7.3 =====
Feb 21 2017 - Submit Date
Feb 24 2017 - App Store Availability

* Initial support for IPv6-only networks.

* Simplified the initial "Loading..." progress screen.

* Tor updated to 0.3.0.3-alpha.

===== 1.7.2 =====
Dec 14 2016 - Submit Date
Dec 15 2016 - App Store Availability

* OpenSSL downgraded to 1.0.2j.

* The OpenSSL version change fixes severe crash issue on all 32bit iOS
  devices. This affected all users of iPhone 5C, iPhone 5, iPhone 4S, iPad
  Mini 1, iPad 4, iPod Touch, and other older iOS devices.

===== 1.7.1 =====
Dec 13 2016 - Submit Date
Dec 13 2016 - App Store Availability

* Adjusted build config to try to fix crash-on-launch issue affecting
  certain devices due to a linking issue.

* Tor updated to 0.2.9.7-rc

===== 1.7.0 =====
Dec 07 2016 - Submit Date
Dec 09 2016 - App Store Availability

* Bridge handling code revised so that the built-in obfs4 bridges get
  updated when Onion Browser updates. Previously you had to re-select
  the built-in bridges to receive new ones.

* Tor, libevent, and OpenSSL are now built using the Tor.framework
  toolkit. See github.com/icepa for more information. Using the
  framework to provide dependencies should improve maintainability
  of the app going forward.

* LICENSE file has been greatly simplified. The Onion Browser 1.7
  series is now made available under the Mozilla Public License 2.0;
  certain unmodified source files may continue to be licensed under
  their original licenses.

* Tor updated to 0.2.9.6-rc

* libevent updated to 2.1.7-rc

* OpenSSL updated to 1.1.0c

===== 1.6.2 =====
Aug 24 2016 - Submit Date
Aug 25 2016 - App Store Availability

* Tor updated to 0.2.8.7, containing several important security fixes:
  https://blog.torproject.org/blog/tor-0287-released-important-fixes

===== 1.6.1 =====
Aug 18 2016 - Submit Date
Aug 19 2016 - App Store Availability

* iObfs rebuilt with golang 1.7 (was 1.7rc3). Fixes crash-on-start when using
  iOS 10 Beta.

===== 1.6.0 =====
Aug 04 2016 - Submit Date
Aug 07 2016 - App Store Availability

* iObfs integration: Onion Browser now supports several "pluggable transport"
  bridge types, including obfs4, meek, and other PT bridge types,
  meaning Onion Browser should connect more successfully in more locations
  that block Tor. Special thanks to The Guardian Project for helping support
  this work. For more info about iObfs (an obfs4proxy iOS port), please visit:
  https://github.com/mtigas/iObfs

* Bridge user interface overhaul: Onion Browser now comes with the same default
  bridges for obfs4, meek-amazon, and meek-azure, as the Tor Browser Bundle and
  Orbot. An easier-to-use bridge page allows users to select the defaults with
  "one click" and without having to manually edit bridge config lines.

* User-agent spoofing strings updated to the latest versions: iOS 9 Safari,
  macOS El Capitan Safari (9.1.1), Tor Browser Bundle 6.0 (Windows 7,
  Firefox 45).

* Tor updated to 0.2.8.6, the first stable release in the 0.2.8 series.
  This version contains several performance and security improvements.
  https://blog.torproject.org/blog/tor-0286-released
  https://gitweb.torproject.org/tor.git/tree/ChangeLog?h=tor-0.2.8.5-rc

* OpenSSL updated to 1.0.2h.
  https://www.openssl.org/news/secadv/20160503.txt
  https://openssl.org/news/changelog.html

* Minimum required iOS version is now iOS 8.2 (was iOS 8.0). Support for iOS
  8.X will continue until Onion Browser 2.0 is released.

===== 1.5.16 =====
Dec 29 2015 - Submit Date
Jan 07 2016 - App Store Availability

* Tor updated to 0.2.7.6.
https://gitweb.torproject.org/tor.git/tree/ChangeLog?h=tor-0.2.7.6

===== 1.5.15 =====
Dec 03 2015 - Submit Date
Dec 10 2015 - App Store Availability

* Tor updated to 0.2.7.5, the first stable release in the new
  Tor 0.2.7 series.
  https://gitweb.torproject.org/tor.git/tree/ChangeLog?h=tor-0.2.7.5

* OpenSSL updated to 1.0.2e, fixing several security issues.
  https://openssl.org/news/secadv/20151203.txt
  https://openssl.org/news/changelog.html

===== 1.5.14 =====
Oct 31 2015 - Submit Date
Nov 08 2015 - App Store Availability

* FIX: Crash when pressing menu button on iPad.

* FIX: GREATLY improve the Onion Browser experience when coming back from
  background. Despite my best efforts to fix the problem, there was a minor
  code bug regarding "appDidBecomeActive" that went unnoticed for a very
  long time. BIG thanks to RustamDz who pointed out the relevant code
  error. (#2 #59)

* Tor updated to 0.2.7.4-rc. This future-proofs Onion Browser a bit
  by supporting stronger Ed25519 encryption keys. The 0.2.7 series also

build-tor.sh
#!/bin/bash
# Builds tor for all five current iPhone targets: iPhoneSimulator-i386,
# iPhoneSimulator-x86_64, iPhoneOS-armv7, iPhoneOS-armv7s, iPhoneOS-arm64.
#
# Copyright 2012-2016 Mike Tigas <mike AT tig DOT as>
#
# Based on "build-libssl.sh" in OpenSSL-for-iPhone by Felix Schulze,
# forked on 2012-02-24. Original license follows:
# Copyright 2010 Felix Schulze. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
###########################################################################
# Choose your tor version and your currently-installed iOS SDK version:
#
#VERSION="0.2.7.6"
VERSION="0.2.8.7"
USERSDKVERSION="10.0"
MINIOSVERSION="8.2"
VERIFYGPG=true

# If you are in a country that blocks access to "torproject.org",
# you will likely get an "Error opening archive" message when building.
# If this happens, delete the tor-*.tar.gz files in the "build/src" directory.
# Then:
#   1. Using one of these URLs, pick a site with "http dist/" or
#      "https dist/". Several are included, in case one is blocked:
#
#     * https://www.torproject.org/getinvolved/mirrors.html.en
#     * https://tor.eff.org/getinvolved/mirrors.html.en
#     * https://www.unicorncloud.org/public/torproject.org/getinvolved/mirrors.html.en
#     * http://beautiful-mind.sky-ip.org/getinvolved/mirrors.html.en
#
#   2. Try accessing the "dist" URL you picked, in your browser. If it
#      does not work, go back to step 1 and try another "http dist/" or
#      "https dist/".
#
#   3. Now paste the dist URL here, making sure that the URL starts with
#      "http://" or "https://" and ends with "/". (The slash at the end
#      is important.)
TOR_DIST_URL="https://dist.torproject.org/"
#TOR_DIST_URL="https://tor.eff.org/dist/"
#TOR_DIST_URL="https://www.unicorncloud.org/public/torproject.org/dist/"
# use this for non-latest versions:
#TOR_DIST_URL="https://archive.torproject.org/tor-package-archive/"
#
# Alternatively, you may download the Tor source ("tor-0.2.5.10.tar.gz" and
# "tor-0.2.5.10.tar.gz.asc") from somewhere else, and place them inside
# the "build/src" directory. Be advised that this may be unsafe, unless
# you already have the GPG keys imported -- see README for details.




###########################################################################
#
# Don't change anything under this line!
#
###########################################################################

# No need to change this since xcode build will only compile in the
# necessary bits from the libraries we create
ARCHS="i386 x86_64 armv7 arm64"

DEVELOPER=`xcode-select -print-path`
#DEVELOPER="/Applications/Xcode.app/Contents/Developer"

# for continuous integration
# https://travis-ci.org/mtigas/iOS-OnionBrowser
if [ "$1" == "--noverify" ]; then
	VERIFYGPG=false
fi
if [ "$2" == "--travis" ]; then
	ARCHS="i386 x86_64"
fi

if [[ ! -z "$TRAVIS" && $TRAVIS ]]; then
	# Travis CI highest available version
	echo "==================== TRAVIS CI ===================="
	SDKVERSION="9.3"
else
	SDKVERSION="$USERSDKVERSION"
fi

cd "`dirname \"$0\"`"
REPOROOT=$(pwd)

# Where we'll end up storing things in the end
OUTPUTDIR="${REPOROOT}/dependencies"
mkdir -p ${OUTPUTDIR}/include
mkdir -p ${OUTPUTDIR}/lib


BUILDDIR="${REPOROOT}/build"

# where we will keep our sources and build from.
SRCDIR="${BUILDDIR}/src"
mkdir -p $SRCDIR
# where we will store intermediary builds
INTERDIR="${BUILDDIR}/built"
mkdir -p $INTERDIR

########################################

cd $SRCDIR

# Exit the script if an error happens
set -e

if [ ! -e "${SRCDIR}/tor-${VERSION}.tar.gz" ]; then
	echo "Downloading tor-${VERSION}.tar.gz"
	#curl -O https://archive.torproject.org/tor-package-archive/tor-${VERSION}.tar.gz
	curl -O ${TOR_DIST_URL}tor-${VERSION}.tar.gz
fi
echo "Using tor-${VERSION}.tar.gz"

# see https://www.torproject.org/docs/verifying-signatures.html.en
# and https://www.torproject.org/docs/signing-keys.html.en
# up to you to set up `gpg` and add keys to your keychain
if $VERIFYGPG; then
	if [ ! -e "${SRCDIR}/tor-${VERSION}.tar.gz.asc" ]; then
		curl -O ${TOR_DIST_URL}tor-${VERSION}.tar.gz.asc
	fi
	echo "Using tor-${VERSION}.tar.gz.asc"
	if out=$(gpg --status-fd 1 --verify "tor-${VERSION}.tar.gz.asc" "tor-${VERSION}.tar.gz" 2>/dev/null) &&
	echo "$out" | grep -qs "^\[GNUPG:\] VALIDSIG"; then
		echo "$out" | egrep "GOODSIG|VALIDSIG"
		echo "Verified GPG signature for source..."
	else
		echo "$out" >&2
		echo "COULD NOT VERIFY PACKAGE SIGNATURE..."
		exit 1
	fi
fi

rm -fr "${SRCDIR}/tor-${VERSION}"
tar zxf tor-${VERSION}.tar.gz -C $SRCDIR
cd "${SRCDIR}/tor-${VERSION}"

####
# Patch to remove the "DisableDebuggerAttachment" ptrace() calls
# that are not allowed in App Store apps
#   cp build/src/tor-0.2.5.8-rc/src/common/compat.c{,.orig}
#   make the edit (see diff)
#   diff -U5 build/src/tor-0.2.5.8-rc/src/common/compat.c{.orig,} > build-patches/tor-ptrace.diff
patch -p3 < ../../../build-patches/tor-ptrace.diff

# Patch to remove "_NSGetEnviron()" call not allowed in App Store
# apps (even fails to compile under iPhoneSDK due to that function
# being undefined)
#   cp build/src/tor-0.2.5.8-rc/src/common/compat.c{.orig,} # after above patch
#   make the edit (see diff)
#   diff -U5 build/src/tor-0.2.5.8-rc/src/common/compat.c{.orig,} > build-patches/tor-nsenviron.diff
patch -p3 < ../../../build-patches/tor-nsenviron.diff

#####
# Collect libz.dylib from the iPhoneSimulator.sdk and iPhoneOS.sdk (already contains armv6 and armv7)
# and compile into a single libz.a file (since that is what tor is looking for)

#lipo -create ${DEVELOPER}/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator${SDKVERSION}.sdk/usr/lib/libz.dylib \
#${DEVELOPER}/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS${SDKVERSION}.sdk/usr/lib/libz.dylib \
#-output ${OUTPUTDIR}/lib/libz.a

cp -R ${DEVELOPER}/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator${SDKVERSION}.sdk/usr/include/zlib* ${OUTPUTDIR}/include/

# Copy ptrace in (since the header is only available on iPhoneSimulator.sdk
# but the compile steps still need it for iPhoneOS)
mkdir -p ${OUTPUTDIR}/include/sys/
cp -R ${DEVELOPER}/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator${SDKVERSION}.sdk/usr/include/sys/ptrace.h ${OUTPUTDIR}/include/sys/

#####
# Now compile tor

cd "${SRCDIR}/tor-${VERSION}"

set +e # don't bail out of bash script if ccache doesn't exist
CCACHE=`which ccache`
if [ $? == "0" ]; then
	echo "Building with ccache: $CCACHE"
	CCACHE="${CCACHE} "
else
	echo "Building without ccache"
	CCACHE=""
fi
set -e # back to regular "bail out on error" mode

export ORIGINALPATH=$PATH

# Keep a copy of `configure` around. In non-64bit arm, we need to
# explicitly disable curve25519-donna-c64, since it causes compile
# error under clang (Xcode 5) for ARM (non-64) archs
export VANILLA_CONFIGURE="${INTERDIR}/tor-${VERSION}.configure"
cp "${SRCDIR}/tor-${VERSION}/configure" "${VANILLA_CONFIGURE}"

for ARCH in ${ARCHS}
do
	if [ "${ARCH}" == "i386" ] || [ "${ARCH}" == "x86_64" ]; then
		PLATFORM="iPhoneSimulator"
		EXTRA_CONFIG=""
	else
		PLATFORM="iPhoneOS"
		EXTRA_CONFIG="--disable-tool-name-check --host=arm-apple-darwin14 --target=arm-apple-darwin14 --disable-gcc-hardening --disable-linker-hardening"
	fi

	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk"
	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include"
	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib"

	export PATH="${DEVELOPER}/Toolchains/XcodeDefault.xctoolchain/usr/bin/:${DEVELOPER}/Platforms/${PLATFORM}.platform/Developer/usr/bin/:${DEVELOPER}/Toolchains/XcodeDefault.xctoolchain/usr/bin:${DEVELOPER}/usr/bin:${ORIGINALPATH}"
	export CC="${CCACHE}`which gcc` -arch ${ARCH} -miphoneos-version-min=${MINIOSVERSION}"

	# (since we're editing configure, make sure to start with an modified
	# copy when building a new arch)
	cp "${VANILLA_CONFIGURE}" "${SRCDIR}/tor-${VERSION}/configure"

	# explicitly disable curve25519-donna-c64, since it causes compile
	# error under clang (Xcode 5) for ARM (non-64) archs
	if [ "${ARCH}" == "armv7" ] || [ "${ARCH}" == "armv7s" ]; then
		sed -ie "s/tor_cv_can_use_curve25519_donna_c64=cross/tor_cv_can_use_curve25519_donna_c64=no/g" "${SRCDIR}/tor-${VERSION}/configure"
		sed -ie "s/tor_cv_can_use_curve25519_donna_c64=yes/tor_cv_can_use_curve25519_donna_c64=no/g" "${SRCDIR}/tor-${VERSION}/configure"
	fi

	./configure ${EXTRA_CONFIG} \
	--prefix="${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk" \
	--enable-static-openssl --enable-static-libevent \
	--with-openssl-dir="${OUTPUTDIR}" \
	--with-libevent-dir="${OUTPUTDIR}" \
	--disable-asciidoc --disable-transparent --disable-threads \
	LDFLAGS="$LDFLAGS -L${OUTPUTDIR}/lib -lz" \
	CFLAGS="$CFLAGS -Os -I${OUTPUTDIR}/include -isysroot ${DEVELOPER}/Platforms/${PLATFORM}.platform/Developer/SDKs/${PLATFORM}${SDKVERSION}.sdk" \
	CPPFLAGS="$CPPFLAGS -I${OUTPUTDIR}/include -isysroot ${DEVELOPER}/Platforms/${PLATFORM}.platform/Developer/SDKs/${PLATFORM}${SDKVERSION}.sdk"

	# Build the application
	make -j$(sysctl hw.ncpu | awk '{print $2}')

	# Don't make install. We actually don't want the tor binary or the
	# documentation, we just want the archives of the compiled sources.
	cp src/common/libor-crypto.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/common/libor-event.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/common/libor.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/common/libcurve25519_donna.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/or/libtor.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/ext/ed25519/ref10/libed25519_ref10.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/ext/ed25519/donna/libed25519_donna.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/trunnel/libor-trunnel.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"
	cp src/ext/keccak-tiny/libkeccak-tiny.a "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/"

	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/common/"
	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/or/"
	mkdir -p "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/tools/"
	cp micro-revision.i "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/"
	cp orconfig.h "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/"
	find src/common -name "*.h" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/common/" \;
	find src/ext -name "*.h" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/" \;
	find src/ext/keccak-tiny -name "*.h" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/" \;
	find src/or -name "*.h" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/or/" \;
	find src/or -name "*.i" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/or/" \;
	find src/tools -name "*.h" -exec cp {} "${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/tools/" \;
	make clean
done

########################################

echo "Build library..."

# These are the libs that comprise tor's internals
OUTPUT_LIBS="libor-crypto.a libor-event.a libor.a libtor.a libcurve25519_donna.a libed25519_ref10.a libed25519_donna.a libor-trunnel.a libkeccak-tiny.a"

for OUTPUT_LIB in ${OUTPUT_LIBS}; do
	INPUT_LIBS=""
	for ARCH in ${ARCHS}; do
		if [ "${ARCH}" == "i386" ] || [ "${ARCH}" == "x86_64" ];
		then
			PLATFORM="iPhoneSimulator"
		else
			PLATFORM="iPhoneOS"
		fi
		INPUT_ARCH_LIB="${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/lib/${OUTPUT_LIB}"
		if [ -e $INPUT_ARCH_LIB ]; then
			INPUT_LIBS="${INPUT_LIBS} ${INPUT_ARCH_LIB}"
		fi
	done
	# Combine the three architectures into a universal library.
	if [ -n "$INPUT_LIBS" ]; then
		lipo -create $INPUT_LIBS \
		-output "${OUTPUTDIR}/lib/${OUTPUT_LIB}"
	else
		echo "$OUTPUT_LIB does not exist, skipping (are the dependencies installed?)"
	fi
done

for ARCH in ${ARCHS}; do
	if [ "${ARCH}" == "i386" ] || [ "${ARCH}" == "x86_64" ];
	then
		PLATFORM="iPhoneSimulator"
	else
		PLATFORM="iPhoneOS"
	fi
	cp -R ${INTERDIR}/${PLATFORM}${SDKVERSION}-${ARCH}.sdk/include/* ${OUTPUTDIR}/include/
	if [ $? == "0" ]; then
		# We only need to copy the headers over once. (So break out of forloop
		# once we get first success.)
		break
	fi
done

mkdir -p ${OUTPUTDIR}/share
cp "${SRCDIR}/tor-${VERSION}/src/config/geoip" ${OUTPUTDIR}/share
cp "${SRCDIR}/tor-${VERSION}/src/config/geoip6" ${OUTPUTDIR}/share

####################

echo "Building done."
echo "Cleaning up..."
# Remove the copy of libz we created. (Our XCode project will use the system
# version.) Ditto with sys/pthread.h.
rm -f ${OUTPUTDIR}/lib/libz.a
rm -f ${OUTPUTDIR}/include/zlib*
rm -fr ${OUTPUTDIR}/include/sys/

# Remove intermediary and source directories
rm -fr ${INTERDIR}
rm -fr "${SRCDIR}/tor-${VERSION}"
echo "Done."
